packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs

```diff
@@ -48,73 +48,85 @@ impl PostgresDatabase {
     /// - [`DatastoreError`], if checking for the [`BaseUri`] failed.
     ///
     /// [`BaseUri`]: crate::types::BaseUri
-    async fn contains_base_uri<'e, E>(executor: E, base_uri: &BaseUri) -> Result<bool, QueryError>
-    where
-        E: Executor<'e, Database = Postgres>,
-    {
-        let (exists,) =
-            sqlx::query_as(r#"SELECT EXISTS(SELECT 1 FROM base_uris WHERE base_uri = $1);"#)
-                .bind(base_uri)
-                .fetch_one(executor)
-                .await
-                .report()
-                .change_context(QueryError)
-                .attach_printable_lazy(|| base_uri.clone())?;
-
-        Ok(exists)
+    async fn contains_base_uri(
+        transaction: &mut Transaction<'_, Postgres>,
+        base_uri: &BaseUri,
+    ) -> Result<bool, QueryError> {
+        Ok(transaction
```

Comment:
It's only because of the poor implementation of the `Executor` trait. We need to use a `Transaction` here because we must not insert the type if any other inserts are failing.

Reply:
My lack of past experience with this is holding me up a bit. Sorry for the work but can you use some more words and explain it in greater detail here? How does this transaction work with some of the other methods? Maybe it'd be helpful to loosely explain what happens with the transaction with`insert_property_type` or `update_property_type` or whatever.

Reply:
Getting rid of the lifetimes is pretty nice as well, but a minor thing

Reply:
Sure. To insert a property type, we want to check, if all references are valid. This is done in `insert_property_type_references`. That function will first read the `VersionId`s for the references. It will fail if a reference does not exist. The version ids are then used to store the reference inside the database. This implies, that the version id of the property type, which we want to insert, is already existing in the database, so inserting the actual type happens before.
Internally, inserting the `property_type` will use transactions (`fn create()`), as multiple things have to be inserted in order to store the blob. The scope of the transaction is now expanded to `insert_property_type_references`, so inserting the property type will fail if the references could not be stored.

Reply:
I don't think I'm getting the information I need here 😁 lets go to a call

Reply:
We cleared that up :) and also, transactions are pain in SQLx